blog/2022/storage-map.adoc

```diff
@@ -0,0 +1,127 @@
+:title: New storage in Keycloak
+:date: 2022-07-27
+:publish: true
+:author: Hynek Mlnařík
+
+Store in Keycloak has known limitations:
+For small deployments, it takes too long to initialize the database and
+start Keycloak. There is no native support for cloud-native deployment.
+Upgrading to a new Keycloak version often means requirement for stop-the-world
+updates with little chance for no downtime upon upgrade.
+
+Keycloak 19 brings in an early preview of the future Keycloak store supporting
+no-downtime upgrades, per-realm storage, and cloud-nativity from its very inception.
+This so caled _map_ storage stays focused at delivering optimal experience and thus
+limits its support to Postgres and CockroachDB databases and Infinispan datastore.
+The early preview is available in Quarkus distribution.
+
+This early preview lacks several features necessary for optimal performance
+and is thus not yet production ready; still we have a plan to deliver those:
+
+- In Keycloak 20, expect improved support especially for CockroachDB;
+  and also a file-based store. Together with another anticipated feature
+  - _tree_ store - it would be possible to combine several storage mechanisms
+  and have e.g. several static client declarations in static files managed
+  in a versioning system like Git, combined with dynamic clients stored in a database.
+  Per-request object caching is on radar for Keycloak 20 too and should result in
+  significantly better performance.
+
+- In Keycloak 21, expect LDAP support, and an offline tool for migrating
+  data from legacy store to new one, and further optimizations and garbage collection
+
+Note that the store used in previous vesions (now called _legacy store_) does
+_not_ go away! Since the new store is not yet production ready, it remains the default
```

Comment:
Eventually it will go away ... do you want to be more specific here?

```suggestion
_not_ go away soon! Since the new store is not yet production ready, it remains the default
```

Reply:
Makes sense, but I don't want to be specific in this blog